File: day/22/p1.js
import input from './input.txt';
import { prototypes } from '../../helper.js';

function calculate(num, times) {
    let t = num;
    for (let i = 0; i < times; i++) {
        const a = (t ^ (t * 64)) % 16777216;
        const b = (a ^ Math.floor(a / 32)) % 16777216;
        const c = (b ^ (b * 2048)) % 16777216;
        t = c;
    }
    return t;
}

function sol(data) {
    const nums = data.trim().split("\n").map(n => parseInt(n, 10));
    const times = 2000;
    let total = 0;
    for (let num of nums) {
        total += calculate(num, times);
    }
    return total;
}

prototypes();
console.log(sol(input));

File: day/22/p1.py
import math

def calculate(num, times):
    t = num
    for i in range(times):
        a = (t ^ (t * 64)) % 16777216
        b = (a ^ (a // 32)) % 16777216
        c = (b ^ (b * 2048)) % 16777216
        t = c
    return t

def sol():
    data = open("input.txt").read()
    nums = map(lambda x: int(x), data.strip().split("\n"))
    times = 2000
    total = 0
    for num in nums:
        ans = calculate(num, times)
        total += ans
    return total

print(sol())